### PR TITLE
feat!: Allow impls on primitive types

### LINKED ISSUE #736: Add impls on primitive types in stdlib
# Problem

We have several functions like `std::array::len(myarray)` which would be more ergonomic to use if they could be accessed with method syntax `myarray.len()`. Other examples include `std::field::to_le_bytes`, to_bits, etc.

# Solution

Allow `impl`s for primitives as long as the impl is defined within noir's stdlib or compiler itself.

### PR DESCRIPTION
**Pull request by jfecher**
# Related issue(s)

Resolves #736 

# Description

Allows impls on primitive types and changes several library functions into methods. This PR does not verify these primitive impls are in the stdlib only as I could not find a good method to do so.

Note that this PR is currently expected to break almost all noir code in its changing of library functions into methods.

## Test additions / changes

Updates many tests to the new method syntax

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.
- [x] This PR requires documentation updates when merged.

# Additional context

What rule should we go by for which library functions should be methods and which (if any) should be free functions? So far, I've started by only making functions that take an argument of the same type into methods. This excludes nullary functions like `std::field::modulus_num_bits` for example.

### CODE REVIEW

**Comment by kevaundray** [comment]
Nice! Will look at this when I get back

**Comment by kevaundray** [comment]
Since this breaks a lot of noir code, we should push a release once this is merged so that we break less newer people's code. 

Alternatively, maybe we could slowly deprécate those functions which are now methods?

**Comment by jfecher** [comment]
I think some discussion on which functions should or should not be methods may also be helpful. Having `.pedersen()` as a method for example makes sense to me and yet still feels rather awkward.

**Comment by jfecher** [comment]
Ah, also worth noting: this PR fixes #618 since for each loops now use `array.len()` internally rather than `dep::std::array::len(array)`. So the problem with the path is no more.

**Comment by kevaundray** [comment]
> Ah, also worth noting: this PR fixes #618 since for each loops now use `array.len()` internally rather than `dep::std::array::len(array)`. So the problem with the path is no more.

Nice, should we kee the issue open and make it more general, since this would fix it for `array::len` but not in general?

**Comment by jfecher** [comment]
> I think for hash functions it feels more natural to have them as functions instead of methods, because unlike .len() there will be other implementations for a hash function that one may want to use, so theres nothing special about the current hash functions that we have in the stdlib, its just that they are the ones we had in our primary backend.

Alright, I'll change these methods back to functions then. Should reduce breakage in user code as a result. What are your thoughts on the `std::field` functions that were changed to methods?

> Nice, should we kee the issue open and make it more general, since this would fix it for array::len but not in general?

I think it is fine to close it. It was an issue because the for-each syntax desugared internally using `dep::std::array::len` and there was no way to change this. Now that it does not do this, there is no other similar construct using a fixed path like this.

**Comment by vezenovm** [comment]
> Alright, I'll change these methods back to functions then. Should reduce breakage in user code as a result. What are your thoughts on the `std::field` functions that were changed to methods?

I think `std::field` functions as methods makes sense as `Field` is a native compiler type. 
```
let x: Field = 5;
let arr = x.to_le_bytes()
```
The snippet above feels natural to me as there is one implementation of this method for the Field type. 

Hashes feel separate from this and it would be awkward to have something like `.pedersen()` as you mentioned above. We don't have a native `Hash` type, so I think keeping a module of different hash functions is good for now. We could bring some type of inheritance structure to the hash functions where all hash functions could implement a `Hash` trait with methods such as `update([u8])` and `finalize() -> [u8]`.

File: crates/noirc_frontend/src/hir/def_collector/dc_crate.rs
Comment (kevaundray) [inline crates/noirc_frontend/src/hir/def_collector/dc_crate.rs:230]: what is this line meant to be doing?

File: crates/noirc_frontend/src/node_interner.rs
Comment (kevaundray) [inline crates/noirc_frontend/src/node_interner.rs:628]: What are the implications of putting function here?

File: crates/noirc_frontend/src/node_interner.rs
Comment (kevaundray) [inline crates/noirc_frontend/src/node_interner.rs:651]: For reference, if we did support adding methods to `Type::Struct` would that add a method to all structs in every Noir codebase?

File: crates/noirc_frontend/src/parser/mod.rs
Comment (kevaundray) [inline crates/noirc_frontend/src/parser/mod.rs:381]: If we changed the method name from `len` to `length`, is there a place that we could define the constant for that, so we don't have a string literal?

File: noir_stdlib/src/array.nr
Comment (kevaundray) [inline noir_stdlib/src/array.nr:11]: noice

File: noir_stdlib/src/array.nr
Comment (kevaundray) [inline noir_stdlib/src/array.nr:56]: yay foreach is back

File: noir_stdlib/src/hash.nr
Comment (kevaundray) [inline noir_stdlib/src/hash.nr:6]: Maybe I'm missing something, why was this change made here?

Comment (kevaundray) [reply noir_stdlib/src/hash.nr:6]: change referring to adding `N`

File: noir_stdlib/src/sha256.nr
Comment (kevaundray) [inline noir_stdlib/src/sha256.nr:103]: Why wasn't `N` added to this like in the sha256 examples above?

File: noir_stdlib/src/sha256.nr
Comment (kevaundray) [inline noir_stdlib/src/sha256.nr:143]: Note: can use += here too

File: noir_stdlib/src/sha256.nr
Comment (kevaundray) [inline noir_stdlib/src/sha256.nr:159]: Note: can use += here -- only noting it as I saw you did it above. Can ignore if you find whats there to be better

File: noir_stdlib/src/sha256.nr
Comment (kevaundray) [inline noir_stdlib/src/sha256.nr:174]: Note: >>= could be done

**Review by kevaundray — COMMENTED** [review]
Left some comments, I like the change.

**Comment by kevaundray** [comment]
> 

+1 to Maxim's comment -- Field looks good already. Need to go back over the specific methods that were changed to methods on Field as there might be some that we only want to be on Field, which _may_ break IntOrField variant. Just something that came OTOH so may not be a problem.

An example would be if we had a method named `shift_right` which would only be for Integers.

**Comment by jfecher** [comment]
> An example would be if we had a method named shift_right which would only be for Integers.

Note that the `IntOrField` variant only controls which functions are _visible_ to integers or fields. The functions themselves can still say they only accept an int or a field (and indeed some methods in std::field do only accept fields). Using these methods on integers you will get a normal type error like you did before.

File: noir_stdlib/src/hash.nr
Comment (jfecher) [reply noir_stdlib/src/hash.nr:6]: I was changing the arrays with anonymous/inferred lengths to ones with explicit lengths in the code that I was touching. It is done to encourage users themselves to use explicit lengths since the inferred ones are still buggy (ref #562). Since users may look into the stdlib to inspect the code or figure out best practices I figured while I was already touching the code I might as well change it.

File: crates/noirc_frontend/src/parser/mod.rs
Comment (jfecher) [reply crates/noirc_frontend/src/parser/mod.rs:381]: Not really, we could move this literal into a constant (there are no other instances of it), but you would still have to change multiple files since you'd have to update array.nr in the noir stdlib to change the len method name there as well.

File: crates/noirc_frontend/src/node_interner.rs
Comment (jfecher) [reply crates/noirc_frontend/src/node_interner.rs:651]: Yes, that'd make a method visible to every struct type. Although in actuality it'd do nothing since we already have struct types and we always handle struct types before falling back to check for these primitive impls if it is not a struct type.

File: crates/noirc_frontend/src/node_interner.rs
Comment (jfecher) [reply crates/noirc_frontend/src/node_interner.rs:628]: It lets us add impls for functions if we really wanted them. In rust for example, you can impl your own traits on function types as well and some libraries make good use of this.

I don't know of any useful methods we could add off the top of my head but one example would be:
```rs
impl<T, U> fn(T) -> U {
    fn call(self, arg: T) -> U {
        self(arg)
    }
}

fn id<T>(x: T) -> T { x }

fn main() {
    constrain id.call(3) == id(3);
}
```

File: crates/noirc_frontend/src/hir/def_collector/dc_crate.rs
Comment (jfecher) [reply crates/noirc_frontend/src/hir/def_collector/dc_crate.rs:230]: Ah, that was leftover from me trying to find a way to restrict primitive impls to only appearing in the stdlib. I could not find a good method in the end since I could not find the crate id of the stdlib, only the local crate.

File: noir_stdlib/src/sha256.nr
Comment (jfecher) [reply noir_stdlib/src/sha256.nr:103]: It was forgotten about since the change was minor. I'll update it

File: crates/noirc_frontend/src/hir/def_collector/dc_crate.rs
Comment (kevaundray) [reply crates/noirc_frontend/src/hir/def_collector/dc_crate.rs:230]: Could we refactor this, If I'm reading it correctly, it says `if true {} else {// do something else}`

File: crates/noirc_frontend/src/node_interner.rs
Comment (kevaundray) [reply crates/noirc_frontend/src/node_interner.rs:628]: Ah makes sense -- Thanks for clarifying ser

File: noir_stdlib/src/hash.nr
Comment (kevaundray) [reply noir_stdlib/src/hash.nr:6]: Yeah that makes sense and thanks for fixing it for the other functions

**Review by kevaundray — COMMENTED** [review]
LGTM -- one nit regarding the if true, apart from that happy to merge

File: crates/noirc_frontend/src/hir/def_collector/dc_crate.rs
Comment (jfecher) [reply crates/noirc_frontend/src/hir/def_collector/dc_crate.rs]: Ah, yes forgot about this block in the last commit. I added a somewhat better check to check that this isn't the local crate at least since we have no way to check if it is the stdlib or not. This should at least prevent most misuses but not all.

**Comment by vezenovm** [comment]
> LGTM -- one nit regarding the if true, apart from that happy to merge

Did we decide whether we are going to merge this before a release or after? CI on the preprocess PR is almost finished

**Review by kevaundray — APPROVED** [review]

File: crates/noirc_frontend/src/hir/def_collector/dc_crate.rs
Comment (kevaundray) [reply crates/noirc_frontend/src/hir/def_collector/dc_crate.rs]: Yeah I think we can open up an issue and tackle this in a later PR. We may need to special case the stdlib in the compiler

**Pull request merged**
